worker/run.py

```diff
@@ -49,56 +51,100 @@ def main():
         args.websocket_base_url = urllib.parse.urlunparse(sync_url)
 
     if args.reload:
+        logging.info("Reload enabled")
         path = Path(__file__).parent
 
-        p, event = run_sync_in_process(args)
-        for _ in watch(path):
+        stop_watching_event = asyncio.Event()
+        p = run_sync_in_process(stop_watching_event, args)
+
+        def watch_filter(change, path: str):
+            # do not reload when models are downloaded
+            return not path.startswith(settings.MODELS_DIR.absolute().as_posix())
+
+        for _ in watch(path, watch_filter=watch_filter, stop_event=stop_watching_event):
             logging.info("Source code change detected, reloading worker")
-            event.set()
+            p.terminate()
             p.join()
-            p, event = run_sync_in_process(args)
+            p = run_sync_in_process(stop_watching_event, args)
 
     else:
-        run_sync(args, Event())
+        run_sync(args)
 
 
-def run_sync_in_process(args):
-    event = Event()
-    p = Process(target=run_sync, args=(args, event))
+def run_sync_in_process(stop_watching_event: asyncio.Event, args):
+    p = Process(target=run_sync, args=(args,))
     p.start()
-    return p, event
+
+    def handle_signal(sig, *args):
+        if p.pid is None:
+            raise Exception("Process has no PID")
+
+        os.kill(p.pid, sig)
+
+        if sig == signal.SIGTERM:
+            stop_watching_event.set()
+        elif sig == signal.SIGUSR1:
+            stop_watching_event.clear()
+
+    signal.signal(signal.SIGTERM, handle_signal)
```

Comment:
why do we do it here again (compared to below?)

Reply:
this is the handling for the reload watcher process, which also is based on pythons multiprocess and therefor differs from the signal handling of the asyncio eventloop of the actual worker process below. The signal handling was a bit tricky due to the watcher process

Reply:
(this also stops the file watching, otherwise it would continue to watch for file changes even after the actual worker process is stopped)

Reply:
can I have comments explaining this?

Reply:
done